Support serving pre-compressed static files

guicamest opened the pull request:
This covers https://github.com/ktorio/ktor/issues/159#issuecomment-376326136

Nowadays, it is quite common to pre-compress static files ( a.k.a assets ) to reduce response time (no need for on the fly compression).

This pull request enables that, serving the pre-compressed version of the file, suppressing compression(if installed) and responding appropriate *Content-Type* and *Content-Encoding*. As brotli is not supported at the time of writing, it also covers serving brotli-compressed files.

For example, if foo.js has been pre-compressed to foo.js.br, a request to foo.js (with appropriate *Accept-Encoding* headers) will result in foo.js.br being served, *Content-Encoding* set to *br* and *Content-Type* set to *application/javascript*.

The same applies to gz files. If preCompressed is called, foo.js.gz will no longer be available @ foo.js.gz (with *Content-Type*: *application/x-compressed*). Instead it will become available @ foo.js with *Content-Encoding* set to *gzip* and *Content-Type* set to *application/javascript*. This also involves suppressing compression.

Example usage:
```
static {
    preCompressed()
    files("static")
}
```

Preferences and configuration of compression algorithms example:
Gzip first:
```
static {
    preCompressed(CompressedFileType.GZIP, CompressedFileType.BROTLI)
    files("static")
}
```

Brotli first:
```
static {
    preCompressed(CompressedFileType.BROTLI, CompressedFileType.GZIP)
    files("static")
}
```

guicamest commented:
@cy6erGn0m @e5l What do you think of this? It'd be great if it makes it into 0.9.4 :) By the way, does it have a date yet?

guicamest commented:
notstale, waiting for feedback from the maintainers

guicamest commented:
Will review conflicts and update the code

guicamest commented:
notstale, waiting for feedback from the maintainers. Will review conflicts and update the code

guicamest commented:
notstale, waiting for feedback from the maintainers. Will review conflicts and update the code asap

guicamest commented:
@cy6erGn0m PR updated with `master` and ready for review

guicamest commented:
@orangy @cy6erGn0m @e5l 
Some feedback from the experience I had (maybe this can be improved):

When running tests from the IDE, the ones using resources failed (StatusPageTest testStatusMappingWithRoutes and testStatusMapping), most likely due to
misconfigured classpath. When I moved the test resources (error404.html and logback-test.xml) manually so that they are in the classpath, they passed.
I used IntelliJ Ultimate 2019.3 and imported using gradle.

When running them from the console on my local machine, they took a really long time (Specially the *BigFile ones).
Also had a time out in io.ktor.tests.websocket.DefaultWebSocketTest.closeByClient with this exception:
org.junit.runners.model.TestTimedOutException: test timed out after 2000 milliseconds

As those tests are not related to the changes introduced in this PR, I assume they are going to pass in the build pipeline.

guicamest commented:
Not stale. Waiting for feedback from the maintainers and reviews

guicamest commented:
PR is not stale, it is waiting for reviews and feedback.
Will update the conflicting text file.

guicamest commented:
PR updated with master and ready for review, again.

guicamest commented:
PR updated with master and ready for review, again.

guicamest commented:
Previous build was failing due to unrelated code (ktor client tests). I just rebased with master and pushed the updated branch.
Keep me posted if you need me to do something else.

emrul commented:
This seems like an excellent patch and I hope it can be approved soon since I would certainly benefit.

guicamest commented:
@hhariri @e5l Thanks for the talk today on the Kotlin Online Event. I have to say I have mixed feelings when I hear you say about contributing and giving feedback. I was really interested in the beginning (more than 2 years ago) and contributed with this pull request and feedback for another issue (filed by Roman) [first comment] (https://github.com/ktorio/ktor/issues/350#issuecomment-413328519) [second comment](https://github.com/ktorio/ktor/issues/350#issuecomment-485112160). Both are still open, with almost no feedback from the team so far.
Today I learned (as suspected) that the project was only being developed and maintained by 2 or 3 developers a long time but now there are more people involved, which hopefully is great news. Is the feedback response from your side going to improve now? I understand there are a lot of things to be done, and almost everything is top-priority, but if contributions and feedback TO the project are important, feedback FROM the project team is needed as well.
I really like Kotlin and I think Ktor has a lot of potential to be a huge player among backend frameworks. Please receive this as a constructive feedback and wish you the best for the current and future roadmap.

hhariri commented:
@guicamest Thanks for the comments and the feedback. Yes, things are improving and will continue to. We'll follow-up with the other comments.

cy6erGn0m reviewed (changes requested):
Looks good in general but please fix formatting and several minor things I've mentioned.

guicamest commented:
> Looks good in general but please fix formatting and several minor things I've mentioned.
Thanks for the code review Sergei, will take a look shortly and update correspondingly

rsinukov commented:
@e5l @guicamest FYI, I changed the base branch to 1.5.0. Since this PR has API change, it can't go to the patch release.

guicamest commented on ktor-server/ktor-server-core/jvm/src/io/ktor/http/content/StaticContent.kt:
```diff
@@ -14,6 +17,33 @@ private const val pathParameterName = "static-content-path-parameter"
+ *
+ * /foo/bar.js.br can be found @ /foo/bar.js
+ *
+ * Appropriate headers will be set and compression will be suppressed if pre-compressed file is found
```
Makes total sense, will document it accordingly.

guicamest commented on ktor-server/ktor-server-core/jvm/src/io/ktor/http/content/StaticContent.kt:
```diff
@@ -85,13 +113,32 @@ fun Route.files(folder: String) = files(File(folder))
  */
 fun Route.files(folder: File) {
     val dir = staticRootFolder.combine(folder)
+    val compressedTypes = staticContentEncodedTypes
```
It is useful for the use-case where some assets are pre-compressed with different algorithms. For example:
```
static
\ -- js (compressed with gz)
\ -- css (compressed with br)
```
can be achieved with:
```
preCompressed(GZIP)
files("js")
preCompressed(BROTLI)
files("css")
```
Does it make sense? @cy6erGn0m 
I agree this needs to be documented

cy6erGn0m replied:
Doesn't it makes more sense to have it structured, like 

```kotlin
preCompressed(GZIP) {
    files("js")
}
preCompressed(BROTLI) {
    files("css")
}
files("videos") // no compressed
```

cy6erGn0m replied:
And if not, we definitely need a test for such case to ensure that BROTLI wins for CSS

guicamest replied:
Hi @cy6erGn0m , regarding 
```
	files("xyz") // not precompressed
	preCompressed(GZIP)
	files("js") // gzip
	preCompressed(BROTLI)
	files("css") // brotli for this and the ones below
```
I added a test (not yet pushed) for it and it is working as expected.

About the dsl
```
	preCompressed(GZIP) {
	    files("js") // GZIP only
	}
```
I see it more of a configuration of the files, unlike enclosing `files` inside a `preCompressed`, which feels like hierarchical. One advantage it has, is that it would allow many `files()` inside a `preCompressed` block.

A possible disadvantage, is the resulting dsl in case another 'configuration' for the `files()` is needed in the future. In case it is also implemented as a block, then the code would look like:
```
	featureX {
		preCompressed(GZIP) {
	    	files("js") // GZIP only
		}
	}
	...
	featureY {
		featureX {
			preCompressed(GZIP) {
	    		files("js") // GZIP only
			}
		}
	}
```
An alternative implementation is having a configuration block for ```files()```. This would result in the following usage:
```
	files("js") {
	    preCompressed(GZIP) // GZIP only
	}
	files("css") {
	    preCompressed(BROTLI) // BROTLI only
	}
	files("videos") // not compressed
```
It would allow future features for ```files``` to be added without changing the dsl:
```
	files("js") {
	    preCompressed(GZIP) // GZIP only
	    featureX()
	}
	files("css") {
	    featureX()
	    featureY = 'abs'
	    preCompressed(BROTLI) // BROTLI only
	}
	files("videos") // not compressed
```
The disadvantage comes when the user has many folders using preCompressed files, it would look like:
```
	files("js") {
	    preCompressed(GZIP) // GZIP only
	}
	files("css") {
		// Again the same config
	    preCompressed(GZIP) // GZIP only
	}
```

What do you think? Should I implement either of those? Or could it be addressed and refactored in a future once a decision on this part of the DSL is made?

cy6erGn0m replied:
This is why nesting "files" in "preCompressed" in better in the last example. 
Regarding the example with featureX and Y: there is a chance of having different config block types but it's unlikely because one wants to have all functions like "file", "files" and other to work inside so both featureX and featureY should have the same `Route` receiver and, therefore, no problem since we are using `Route` to extend as well.

The disadvantage of nesting comparing to the flat model is that it intoduces nesting that makes a little bit more "noize". But I believe it looks more obvious, eliminates questions and possible confusion.

guicamest replied:
Thanks for the feedback @cy6erGn0m , I'll add the requested changes

guicamest replied:
Updated with the requested changes

guicamest commented on ktor-server/ktor-server-core/jvm/src/io/ktor/http/content/StaticContent.kt:
```diff
@@ -85,13 +113,32 @@ fun Route.files(folder: String) = files(File(folder))
+    val bestCompressionFit = file.bestCompressionFit(request.acceptEncodingItems(), compressedTypes)
+    bestCompressionFit?.apply {
+        attributes.put(Compression.SuppressionAttribute, true)
+        response.header(HttpHeaders.ContentEncoding, encoding)
```
I found `wrapHeaders` in the ktor-client module, so I cannot use it here. Is it planned to be available in the server module as well? If not, how would you recommend to solve this?
1. Move `Content.kt` from the client package into a shared package
2. Copy the Content.kt from the client package into the server package
3. Create `private fun OutgoingContent.ReadChannelContent.wrapHeaders(block: (Headers) -> Headers): OutgoingContent.ReadChannelContent` and use it in `StaticContent` privately
4. `respond`
```object : OutgoingContent.ReadChannelContent() {
        override val contentLength: Long? get() = this@wrapHeaders.contentLength
        override val contentType: ContentType? get() = this@wrapHeaders.contentType
        override val status: HttpStatusCode? get() = this@wrapHeaders.status

        override val headers: Headers = block(this@wrapHeaders.headers)

        override fun readFrom(): ByteReadChannel = this@wrapHeaders.readFrom()

        override fun readFrom(range: LongRange): ByteReadChannel = this@wrapHeaders.readFrom(range)
    }```

cy6erGn0m commented on ktor-server/ktor-server-core/jvm/src/io/ktor/http/content/StaticContent.kt:
```diff
@@ -85,15 +126,55 @@ public fun Route.files(folder: String): Unit = files(File(folder))
+    override fun readFrom() = original.readFrom()
+    override fun readFrom(range: LongRange) = original.readFrom(range)
+    override val headers by lazy(LazyThreadSafetyMode.NONE) {
+        if( encoding != null ) {
```
formatting

```suggestion
        if (encoding != null) {
```

cy6erGn0m commented on ktor-server/ktor-server-core/jvm/src/io/ktor/http/content/StaticContent.kt:
```diff
@@ -14,6 +17,46 @@ private const val pathParameterName = "static-content-path-parameter"
+public fun Route.preCompressed(vararg types: CompressedFileType = CompressedFileType.values()){
+    if (types.isNotEmpty()) {
+        attributes.put(compressedKey, types.asList())
+    }else
```
```suggestion
    } else {
```

cy6erGn0m commented on ktor-server/ktor-server-tests/jvm/test/io/ktor/tests/server/features/StaticContentTest.kt:
```diff
@@ -155,6 +155,119 @@ class StaticContentTest {
+            }
+        }
+
+        handleRequest(HttpMethod.Get, "/${temp.nameWithoutExtension}"){
```
```suggestion
        handleRequest(HttpMethod.Get, "/${temp.nameWithoutExtension}") {
```

cy6erGn0m reviewed (changes requested):
One more iteration

cy6erGn0m commented:
ℹ️  This is the oldest open PR as far as I can see. Congratulations for all 😆

cy6erGn0m commented:
@guicamest you got too many irrelevant commits. You interactive rebase and use drop for every alien commit

guicamest commented:
> 
> 
> @guicamest you got too many irrelevant commits. You interactive rebase and use drop for every alien commit
Fixed it. I also added a test that verifies that the order has an effect on the resolution of the assets.

Is there any update on the dsl change discussion?

cy6erGn0m commented on ktor-server/ktor-server-tests/jvm/test/io/ktor/tests/server/features/StaticContentTest.kt:
```diff
@@ -154,6 +154,164 @@ class StaticContentTest {
+            assertEquals("br", result.response.headers[HttpHeaders.ContentEncoding].orEmpty())
+        }
+    }
+
```
Could you please add a test with preCompressed nesting to ensure that it's working properly? This is the last change request, I believe :)

guicamest replied:
Hi @cy6erGn0m , sorry for the delay, I actually wrote the test but never pushed it :(

Thanks for taking care of it and merging, it feels good to be able to contribute to Ktor!

cy6erGn0m commented:
Added a test myself and merged manually. Thanks, @guicamest for your long work and everyone for patience, and happy ktor coding!

cy6erGn0m commented:
Reference ticket is [KTOR-1447 Support for precompresed files](https://youtrack.jetbrains.com/issue/KTOR-1447)
